### PR TITLE
Gh 485 serialisers null handling

### PR DESCRIPTION
**Pull request by t616178**

### CODE REVIEW

**Comment by t616178** [comment]
This probably needs some more work/testing. I'm having issues building locally, so wanted to start the process off on Travis to see if the same issues are manifest in the CI environment.

**Comment by t616178** [comment]
There are a couple of consequences to the changes that I've made here, so I'm happy to take suggestions if they cause more problems than they solve:

* Null handling behaviour is pushed down from the Accumulo Element converter to the individual serialiser classes. This is to provide enough flexibility to be able to discriminate between null values and empty (or default) values. An AbstractSerialiser class has been added which contains default handling for deserialising null values and for serialising empty values.
* Any Graphs which are created with a visibility property will always result in the visibility property being set on any elements retrieved from that Graph, even if the visibility contains an empty value.
* Element properties have been modified so that it is not permitted to create a property with a null key or a null value.

# Migration required

The format of the Serialiser classes has changed. The serialisers are now generic classes, with the generic parameter being the type of object serialised and the target type for deserialisation.

The Serialiser classes are now also responsible for handling nulls. Default behaviour is provided by the AbstractSerialiser class, which should be extended. The default behaviour is to serialise null values as empty byte arrays and to deserialise empty byte arrays to null values again. In combination with the configuration for the Jackson ObjectMapper used, these null values will not be serialised when written as JSON.

File: accumulo-store/src/main/java/gaffer/accumulostore/key/core/AbstractCoreKeyAccumuloElementConverter.java [hunk 1]
```diff
@@ -272,8 +273,15 @@ public Properties getPropertiesFromColumnVisibility(final String group, final by
             if (null != propertyDef) {
                 final Serialisation serialiser = propertyDef.getSerialiser();
                 try {
-                    properties.put(schema.getVisibilityProperty(),
-                            serialiser.deserialise(columnVisibility));
+                    if (columnVisibility == null || columnVisibility.length == 0) {
+                        if (serialiser.deserialiseEmptyBytes() != null) {
```
Comment (p013570) [inline accumulo-store/src/main/java/gaffer/accumulostore/key/core/AbstractCoreKeyAccumuloElementConverter.java]: Perhaps store the response from serialiser.deserialiseEmptyBytes() so we don't have to call it twice.

File: accumulo-store/src/test/java/gaffer/accumulostore/integration/AccumuloVisibilityIT.java [hunk 2]
```diff
@@ -0,0 +1,430 @@
+package gaffer.accumulostore.integration;
```
Comment (p013570) [inline accumulo-store/src/test/java/gaffer/accumulostore/integration/AccumuloVisibilityIT.java]: These tests are really helpful and I think they would be useful for any Store implementation. Could we move them into the gaffer-integration-tests module and set the class to extend AbstractStoreIT. We could then reuse them for all Stores.

Comment (p013570) [inline accumulo-store/src/test/java/gaffer/accumulostore/integration/AccumuloVisibilityIT.java]: @t616178 did you manage to make this change?

Comment (t616178) [inline accumulo-store/src/test/java/gaffer/accumulostore/integration/AccumuloVisibilityIT.java]: Yes, this has been implemented now.

**Pull request merged**